op-e2e/actions/l1_miner.go

```diff
@@ -0,0 +1,142 @@
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+// L1Miner wraps a L1Replica with instrumented block building ability.
+type L1Miner struct {
+	L1Replica
+
+	// L1 block building data
+	l1BuildingHeader *types.Header             // block header that we add txs to for block building
+	l1BuildingState  *state.StateDB            // state used for block building
+	l1GasPool        *core.GasPool             // track gas used of ongoing building
+	pendingIndices   map[common.Address]uint64 // per account, how many txs from the pool were already included in the block, since the pool is lagging behind block mining.
+	l1Transactions   []*types.Transaction      // collects txs that were successfully included into current block build
+	l1Receipts       []*types.Receipt          // collect receipts of ongoing building
+	l1Building       bool
+	l1TxFailed       []*types.Transaction // log of failed transactions which could not be included
+}
+
+// NewL1Miner creates a new L1Replica that can also build blocks.
+func NewL1Miner(log log.Logger, genesis *core.Genesis) *L1Miner {
+	rep := NewL1Replica(log, genesis)
+	return &L1Miner{
+		L1Replica: *rep,
+	}
+}
+
+// ActL1StartBlock returns an action to build a new L1 block on top of the head block,
+// with timeDelta added to the head block time.
+func (s *L1Miner) ActL1StartBlock(timeDelta uint64) Action {
+	return func(t Testing) {
+		if s.l1Building {
+			t.InvalidAction("not valid if we already started building a block")
+		}
+		if timeDelta == 0 {
+			t.Fatalf("invalid time delta: %d", timeDelta)
+		}
+
+		parent := s.l1Chain.CurrentHeader()
+		parentHash := parent.Hash()
+		statedb, err := state.New(parent.Root, state.NewDatabase(s.l1Database), nil)
+		if err != nil {
+			t.Fatalf("failed to init state db around block %s (state %s): %w", parentHash, parent.Root, err)
+		}
+		header := &types.Header{
+			ParentHash: parentHash,
+			Coinbase:   parent.Coinbase,
+			Difficulty: common.Big0,
+			Number:     new(big.Int).Add(parent.Number, common.Big1),
+			GasLimit:   parent.GasLimit,
+			Time:       parent.Time + timeDelta,
+			Extra:      []byte("L1 was here"),
+			MixDigest:  common.Hash{}, // TODO: maybe randomize this (prev-randao value)
+		}
+		if s.l1Cfg.Config.IsLondon(header.Number) {
+			header.BaseFee = misc.CalcBaseFee(s.l1Cfg.Config, parent)
+			// At the transition, double the gas limit so the gas target is equal to the old gas limit.
+			if !s.l1Cfg.Config.IsLondon(parent.Number) {
+				header.GasLimit = parent.GasLimit * params.ElasticityMultiplier
+			}
+		}
+
+		s.l1Building = true
+		s.l1BuildingHeader = header
+		s.l1BuildingState = statedb
+		s.l1Receipts = make([]*types.Receipt, 0)
+		s.l1Transactions = make([]*types.Transaction, 0)
+		s.pendingIndices = make(map[common.Address]uint64)
+
+		s.l1GasPool = new(core.GasPool).AddGas(header.GasLimit)
+	}
+}
+
+// ActL1IncludeTx includes the next tx from L1 tx pool from the given account
```

Comment:
This includes a single tx from a particular sender that is in the mempool. Is `TxPool().ContentFrom` deterministic?

Reply:
The mempool order per account is deterministic, it keeps the txs with the lowest nonce value around, in order. And separates pending txs from queued ones (when it can't be processed due to nonce or fee etc.)